### PR TITLE
[export] Serialize metadata

### PR DESCRIPTION
**Pull request by angelayi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103274

### CODE REVIEW

File: torch/_export/serde/serialize.py [hunk 1]
```diff
@@ -194,7 +207,47 @@ def deserialize_metadata(metadata) -> Dict[str, str]:
...
+            key = kv[:key_idx]
+            assert kv[key_idx + 1] == "("
+            assert kv[-1] == ")"
+            values = kv[key_idx + 2: -1].split(",")
+            assert len(values) == 2
+            module = deserialize_meta_func(values[1])
+            nn_module_stack[key] = (values[0], module)
+        ret["nn_module_stack"] = nn_module_stack
+
+    if source_fn_str := metadata.get("source_fn"):
```
Comment (zhxchen17) [inline torch/_export/serde/serialize.py:245]: who needs source_fn to be serialized?

File: torch/_export/serde/serialize.py [hunk 2]
```diff
@@ -194,7 +207,47 @@ def deserialize_metadata(metadata) -> Dict[str, str]:
...
+
+        target = module
+        for name in serialized_target_names:
+            if not hasattr(target, name):
+                return serialized_target
+            else:
+                target = getattr(target, name)
+        return target
+
+    if nn_module_stack_str := metadata.get("nn_module_stack"):
```
Comment (zhxchen17) [inline torch/_export/serde/serialize.py:230]: what's the difference between nn_module_stack and module_fqn?

File: torch/_export/serde/serialize.py (hunk 1 above)
Comment (angelayi) [reply torch/_export/serde/serialize.py:245]: The source_fn is used to partition modules (torch.nn.linear) and functional modules (torch.nn.F.linear). It's almost like nn_module_stack but also works for functions that this node was decomposed from.

File: torch/_export/serde/serialize.py (hunk 2 above)
Comment (angelayi) [reply torch/_export/serde/serialize.py:230]: module_fqn seems like a sigmoid specific thing: https://fburl.com/code/uafys81f

File: torch/_export/serde/serialize.py (hunk 1 above)
Comment (avikchaudhuri) [reply torch/_export/serde/serialize.py:245]: Curious, is this just the fully qualified name of the function? Or something more structured that has been `str()`-ified?

File: torch/_export/serde/serialize.py (hunk 2 above)
Comment (avikchaudhuri) [reply torch/_export/serde/serialize.py:230]: Mmm, worth standardizing these extra fields right now, maybe it's dupe / derivable from something else?

Comment (avikchaudhuri) [reply torch/_export/serde/serialize.py:230]: Looking at the line, it looks like `module_fqn` is derivable from the stack. Not clear whether this is the "current" (innermost) module or the "main" (outermost) module?

File: torch/_export/serde/serialize.py (hunk 1 above)
Comment (angelayi) [reply torch/_export/serde/serialize.py:245]: It's a tuple containing the fully qualified name of the node that was inserted before lowering to aten graph, and the function or module it was before decomposition. The function could also be a string name 😅

**Comment by zhxchen17** [comment]
talked offline, let's keep "source_fn" and "nn_module_stack" and remove "module_fqn" for now.

File: torch/_export/serde/serialize.py [hunk 3]
```diff
@@ -178,27 +178,68 @@ def serialize_metadata(node: torch.fx.Node) -> Dict[str, str]:
...
-        ret["module_fqn"] = module_fqn
-    # TODO(angelayi) add nn_module_stack and source_fn
+
+    if nn_module_stack := node.meta.get("nn_module_stack"):
+        # Serialize to "fx_node_name:(orig_ref,type_str)"
+        nn_module_list = [
+            f"{k}:({v[0]},{serialize_operator(v[1])})"
+            for k, v in nn_module_stack.items()
+        ]
+        ret["nn_module_stack"] = ";".join(nn_module_list)
```
Comment (zhxchen17) [inline torch/_export/serde/serialize.py:188]: On the serializer side, I feel we should always save these metadata. If they're absent, then maybe we should error.

**Review by zhxchen17 — APPROVED** [review]

File: torch/_export/serde/serialize.py (hunk 3 above)
Comment (angelayi) [reply torch/_export/serde/serialize.py:188]: hm, I'm not sure if we should error because some nodes (like assert) dont have meta. Maybe we can warning?

Comment (angelayi) [reply torch/_export/serde/serialize.py:188]: Actually I thought ab it and we can't error -- because not every node has an nn_module_stack if it's not in a module.
